### PR TITLE
DONT MERGE: Adding caller runs and work-stealing

### PR DESCRIPTION
**Pull request by pveentjer**
Performance in Hazelcast is partly dominated by the threading design: when an operation needs to be processed, it is handed over to the correct partition-queue where the partition-thread picks it up and eventually the response is send back using a wait/notifyAll.

It is a lot cheaper if a thread can execute its own operations, this skips this expensive mechanism and reduces the pressure in the partition-threads so that more remote operations can be executed. 

This is implemented using a different OperationScheduler implementation, so changes to the system are quite minimal. In the current OperationScheduler there is a single queue per operation-thread that is shared between different partitions. 

To steal work from this queue is very difficult since this queue contains an interleaved number of operations from different partitions. The new WorkQueue can be seen as a stack of PartitionQueues. Stealing becomes a lot easier since the whole PartitionQueue can be locked in 1 atomic action. When a PartitionQueue is stolen, it isn't removed from the WorkQueue; the partition thread will just ignore it. 

Some performance numbers

----------------------------------------------
only local calls
----------------------------------------------

class=com.hazelcast.stabilizer.tests.concurrent.atomiclong.AtomicLongTest
countersLength = 1000
threadCount = 40
keyLocality = local
writePercentage = 0

before
3,604,602.61 ops/s

after
17,590,757.57 ops/s

that is a 388 % performance improvement.

----------------------------------------------
random calls
----------------------------------------------

class=com.hazelcast.stabilizer.tests.concurrent.atomiclong.AtomicLongTest
countersLength = 1000
threadCount = 40
keyLocality = random
writePercentage = 0

before
677,200.00 ops/second

after
798,601.00 ops/s

that is a 18% performance improvement. Although 18% doesn't sound like a lot, we put in huge efforts to get a few percent. The nice thing about this optimization is that can speed any short operation in the system, so it isn't bound to a particular data-structure.

I'm sure that we can get at least another 100% performance improvements for local calls since a lot of debris is still on the way; for  example when a response is send back to the thread executing the operation (and executing its own operation) there is no need to do a wait/notify to signal a response.

### CODE REVIEW

**Comment by pveentjer** [comment]
Small update.

The following test is very synthetic, so probably not something you will find in production.

If a single thread on a single member does a read on an IAtomicLong (cheapest structure we have), then without the optimization:

51,036.67 ops/s

With the optimization:

1,316,276.67 ops/s

That is almost 25 times faster. This shows how much impact the threading model has.

**Comment by bwzhang2011** [comment]
@pveentjer , thanks for work-steal brought in.  how about some further test e.g for IQueue or ITopic or IMap (we care about that may be could fork your version for testing)or some test in stabilizer to leave us more message.

**Comment by pveentjer** [comment]
ITopics are not going to benefit from this since they don't rely on operation-threads, but on event threads. We'll provide more tests. I hope this new executor can be included as experimental feature in 3.5 that needs to be enabled explicitly.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 1]
```diff
@@ -67,6 +67,12 @@
      * partition operations will queue behind other operations of different partitions. The default is 4.
      */
     public static final String PROP_PARTITION_OPERATION_THREAD_COUNT = "hazelcast.operation.thread.count";
+    public static final String PROP_PROGRESSIVE_SCHEDULER_ENABLED = "hazelcast.operation.progressivescheduler.enabled";
+
+    //public static final String PROP_PROGRESSIVE_SCHEDULER_ENABLED = "hazelcast.kickass.scheduler.enabled";
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: this doesn't belong here

File: hazelcast/src/main/java/com/hazelcast/spi/impl/BasicOperationService.java [hunk 2]
```diff
@@ -169,6 +161,27 @@
         startCleanupThread();
     }
 
+    private OperationExecutor initOperationExecutor() {
+        return new ProgressiveOperationExecutor(
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/BasicOperationService.java]: the progressive executor should be used only when the group property is explicitly enabled, shouldn't it?

File: hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/progressive/PartitionIdScanThread.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import com.hazelcast.partition.InternalPartition;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.util.executor.HazelcastManagedThread;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class PartitionIdScanThread extends HazelcastManagedThread {
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/progressive/PartitionIdScanThread.java]: missing JavaDoc? what's a responsibility of this class?

**Comment by jerrinot** [comment]
there are checkstyle issues.
 
moreover I think it deserves a proper documentation before it's merged - it's relative big architectural change.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/BasicOperationService.java (hunk 2 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/BasicOperationService.java]: For the time being it is enabled by default so I can run all the unit tests and see if something breaks. Will be reverted when I make the final PR.

**Comment by pveentjer** [comment]
I'll close it for now. 

The engine is plugable and not enabled by default. Although documentation is needed and perhaps the whole formal process, it isn't used when a user has not explicitly configured it.

**Comment by bwzhang2011** [comment]
@pveentjer, such issue merged into master branch ?

**Comment by bwzhang2011** [comment]
@pveentjer, I noticed that the E-A doesn't include such feature and as you commented it would be run in plugable, so I wonder such progress.

**Pull request closed without merging**